scripts/dependencies.py

```diff
@@ -0,0 +1,255 @@
+#!/usr/bin/env python3
+
+import argparse
+from copy import deepcopy
+from operator import eq, itemgetter, ge, gt, le, lt, ne
+import os
+import re
+import sys
+from tempfile import TemporaryDirectory
+
+import git
+from schema import SchemaError
+import yaml
+
+from components import load_component, load_data, save_data
+from lib.schemata import (
+    component_metadata_schema, component_requirements_schema, component_schema,
+    branch_constraint_regex, branch_constraints_schema,
+    version_constraint_regex,
+)
+
+METADATA_FILENAME = "component_metadata.yml"
+REQUIREMENTS_FILENAME = "component_requirements.yml"
+
+
+def load_metadata(directory):
+    filepath = os.path.join(directory, METADATA_FILENAME)
+
+    return load_data(filepath, component_metadata_schema)
+
+
+def save_metadata(metadata, directory):
+    filepath = os.path.join(directory, METADATA_FILENAME)
+
+    save_data(filepath, metadata, component_metadata_schema)
+
+
+def load_requirements(directory):
+    filepath = os.path.join(directory, REQUIREMENTS_FILENAME)
+
+    return load_data(filepath, component_requirements_schema)
+
+
+def save_requirements(requirements, directory):
+    filepath = os.path.join(directory, REQUIREMENTS_FILENAME)
+
+    save_data(filepath, requirements, component_requirements_schema)
+
+
+def set_dependency(existing, name, constraints=None):
+    if existing:
+        dependencies = deepcopy(existing)
+    else:
+        dependencies = {"dependencies": []}
+
+    for dependency in dependencies["dependencies"]:
+        if dependency["name"] == name:
+            dependency["constraints"] = constraints
+            break
+    else:
+        dependencies["dependencies"].append(
+            {
+                "name": name,
+                "constraints": constraints,
+            }
+        )
+
+    return component_metadata_schema.validate(dependencies)
+
+
+def build_constraint_checker(constraints):
+    op_map = {
+        "==": eq,
+        "!=": ne,
+        ">=": ge,
+        "<=": le,
+        ">": gt,
+        "<": lt,
+    }
+
+    def partial_flip(fn, y):
+        def inner(x):
+            return fn(x, y)
+        return inner
+
+    checks = []
+    for constraint in constraints:
+        constraint_match = re.match(version_constraint_regex, constraint)
+        checks.append(
+            partial_flip(
+                op_map[constraint_match.group("comparison_operator")],
+                constraint_match.group("version"),
+            )
+        )
+
+    return lambda v: all(c(v) for c in checks)
+
+
+def requirement_from_version_constraints(component, constraints):
+        meets_constraints = build_constraint_checker(constraints)
+        versions = (
+            version
+            for series in component["releases"]
+            for version in series["versions"]
+        )
+        for version in versions:
+            if meets_constraints(version["version"]):
```

Comment:
How should constraints be specified?  If I have a component with two versions (r1.0.0, r1.1.0) and try to update requirements on a component that uses that dependent component, it seems to select the first version of the dependent component that it finds rather than finding the newest version that satisfies the constraint.